File: ios/Chitra/Data/DebugHooks.swift
import Foundation

/**
 Debug-build-only launch environment so a script can configure the app
 without driving the UI — the iOS counterpart of the Android client's
 `am start --es` extras:

     SIMCTL_CHILD_CHITRA_SERVER_URL=http://127.0.0.1:8765 \
     SIMCTL_CHILD_CHITRA_ROUTE=settings \
     SIMCTL_CHILD_CHITRA_FILTER=uploads \
     SIMCTL_CHILD_CHITRA_BACKUP_ENABLE=1 \
     SIMCTL_CHILD_CHITRA_BACKUP_ALBUMS=Recents,Screenshots \
       xcrun simctl launch --terminate-running-process <device> com.buildapp.photos

 (simctl passes any SIMCTL_CHILD_-prefixed variable through to the app; in
 Xcode they go in the scheme's Run > Arguments > Environment Variables.)

 A release build ignores all of it.
 */
enum DebugHooks {
    /// Filter the library should open on, and where to start: a tab
    /// ("memories", "albums", "search") or a screen pushed inside one
    /// ("settings", "people", "map").
    static var initialFilter: Filter?
    static var initialRoute: String?

    static func apply() {
        #if DEBUG
        let env = ProcessInfo.processInfo.environment
        let settings = SettingsStore.shared

        if let url = env["CHITRA_SERVER_URL"], !url.isEmpty { settings.setServerURL(url) }
        if let filter = env["CHITRA_FILTER"] { initialFilter = Filter(rawValue: filter.lowercased()) }
        if let route = env["CHITRA_ROUTE"] { initialRoute = route.lowercased() }
        if let videos = env["CHITRA_BACKUP_VIDEOS"] { settings.setBackupVideos(flag(videos)) }
        if let wifi = env["CHITRA_BACKUP_WIFI_ONLY"] { settings.setBackupWifiOnly(flag(wifi)) }

        if let names = env["CHITRA_BACKUP_ALBUMS"] {
            let wanted = Set(names.split(separator: ",")
                .map { $0.trimmingCharacters(in: .whitespaces).lowercased() }
                .filter { !$0.isEmpty })
            let ids = DeviceMedia.albums().filter { wanted.contains($0.name.lowercased()) }.map(\.id)
            settings.setBackupAlbums(Set(ids))
        }
        if let enable = env["CHITRA_BACKUP_ENABLE"] {
            settings.setBackupEnabled(flag(enable))
        }
        // Wipe the ledger before starting a run, or the service can read it
        // first and find nothing to send.
        if let clear = env["CHITRA_LEDGER_CLEAR"], flag(clear) {
            UploadLedger.shared.clear(server: settings.serverURL)
        }
        #endif
    }

    private static func flag(_ value: String) -> Bool {
        ["1", "true", "yes", "on"].contains(value.lowercased())
    }
}

File: android/app/src/main/java/com/buildapp/photos/data/DebugHooks.kt
package com.buildapp.photos.data

import android.content.Context
import android.content.Intent
import android.content.pm.ApplicationInfo
import kotlinx.coroutines.flow.first

/**
 * Debug-build-only launch extras so the emulator E2E script can configure
 * the app without driving the UI:
 *
 *   am start -n com.buildapp.photos/.MainActivity \
 *      --es server_url http://10.0.2.2:8765 --ez backup_enable true \
 *      --es backup_buckets Camera,Screenshots --ez backup_videos true \
 *      --ez backup_wifi_only false --ez backup_now true --es filter uploads
 *
 * A release build ignores all of it.
 */
object DebugHooks {
    @Volatile var initialFilter: String? = null

    fun isDebuggable(ctx: Context) =
        (ctx.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0

    suspend fun apply(ctx: Context, intent: Intent?) {
        if (intent == null || !isDebuggable(ctx)) return
        val settings = SettingsRepository(ctx)
        intent.getStringExtra("server_url")?.let { settings.setServerUrl(it) }
        // Sign in (accounts on the server): login_name + login_password.
        intent.getStringExtra("login_name")?.let { name ->
            val pw = intent.getStringExtra("login_password") ?: ""
            runCatching {
                val r = com.buildapp.photos.api.PhotoApi.create(settings.serverUrl.first()).login(
                    com.buildapp.photos.api.LoginBody(name, pw, device = "e2e"))
                settings.setSession(r.token, r.user.name)
            }
        }
        intent.getStringExtra("filter")?.let { initialFilter = it }
        if (intent.hasExtra("backup_videos")) settings.setBackupVideos(intent.getBooleanExtra("backup_videos", true))
        if (intent.hasExtra("backup_wifi_only")) settings.setBackupWifiOnly(intent.getBooleanExtra("backup_wifi_only", true))
        intent.getStringExtra("backup_buckets")?.let { names ->
            val wanted = names.split(",").map { it.trim().lowercase() }.filter { it.isNotEmpty() }.toSet()
            val ids = DeviceMedia.buckets(ctx.contentResolver)
                .filter { it.name.lowercase() in wanted }.map { it.id }.toSet()
            settings.setBackupBuckets(ids)
        }
        if (intent.hasExtra("backup_enable")) {
            val on = intent.getBooleanExtra("backup_enable", false)
            settings.setBackupEnabled(on)
            if (on) BackupWorker.schedule(ctx, settings.backup.first().wifiOnly) else BackupWorker.cancel(ctx)
        }
        // Wipe the ledger before starting a run, or the worker can read it
        // first and find nothing to send.
        if (intent.getBooleanExtra("ledger_clear", false)) UploadLedger.get(ctx).clear(settings.serverUrl.first())
        if (intent.getBooleanExtra("backup_now", false)) BackupWorker.runNow(ctx)
    }
}
